File: problems/2583_kth_largest_sum_in_a_binary_tree/solution.py
from typing import Optional
from shared.python.data_structures import TreeNode


class Solution:
    def kthLargestLevelSum(self, root: Optional[TreeNode], k: int) -> int:
        from collections import deque
        sums = []
        q = deque([root])
        while q:
            level_sum = 0
            for _ in range(len(q)):
                node = q.popleft()
                level_sum += node.val
                if node.left: q.append(node.left)
                if node.right: q.append(node.right)
            sums.append(level_sum)
        if k > len(sums): return -1
        sums.sort(reverse=True)
        return sums[k - 1]

File: problems/2583_kth_largest_sum_in_a_binary_tree/Solution.java
import java.util.*;

class Solution {
    public long kthLargestLevelSum(TreeNode root, int k) {
        List<Long> sums = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        while (!q.isEmpty()) {
            long s = 0; int sz = q.size();
            for (int i = 0; i < sz; i++) {
                TreeNode n = q.poll(); s += n.val;
                if (n.left != null) q.offer(n.left);
                if (n.right != null) q.offer(n.right);
            }
            sums.add(s);
        }
        if (k > sums.size()) return -1;
        Collections.sort(sums, Collections.reverseOrder());
        return sums.get(k-1);
    }
}
